File: src/HuuDucTruongVu/File Nộp Bài/Bai1a_DSA_C++/CircularLinkedList/CircularLinkedList.cpp
#include <iostream>
#include <cmath>
#include <list>

using namespace std;

//
struct CircularDoublyLinkedList {
    struct Node {
        int info;
        Node* pNext;
        Node* pPrev;
    };

    Node* pHead;

    CircularDoublyLinkedList() : pHead(nullptr) {}

    bool IsEmpty() {
        return pHead == nullptr;
    }

    void AddFirst(int value) {
        Node* newNode = new Node();
        newNode->info = value;

        if (IsEmpty()) {
            newNode->pNext = newNode;
            newNode->pPrev = newNode;
            pHead = newNode;
        }
        else {
            Node* tail = pHead->pPrev;
            newNode->pNext = pHead;
            newNode->pPrev = tail;
            tail->pNext = newNode;
            pHead->pPrev = newNode;
            pHead = newNode;
        }
    }

    void AddLast(int value) {
        Node* newNode = new Node();
        newNode->info = value;

        if (IsEmpty()) {
            AddFirst(value);
        }
        else {
            Node* tail = pHead->pPrev;
            newNode->pNext = pHead;
            newNode->pPrev = tail;
            tail->pNext = newNode;
            pHead->pPrev = newNode;
        }
    }

    void AddAfter(int value, Node* node) {
        if (node == nullptr) {
            cout << "Cannot add after null." << endl;
            return;
        }

        Node* newNode = new Node();
        newNode->info = value;

        newNode->pNext = node->pNext;
        newNode->pPrev = node;

        if (node->pNext) { 
            node->pNext->pPrev = newNode;
        }
        node->pNext = newNode;
    }
    
    void RemoveFirst() {
        if (IsEmpty()) {
            cout << "The List is empty." << endl;
            return;
        }

        if (pHead == pHead->pNext) {
            delete pHead;
            pHead = nullptr;
            return;
        }

        Node* tail = pHead->pPrev;
        Node* originalHead = pHead;
        Node* newHead = pHead->pNext;

        pHead = newHead;
        pHead->pPrev = tail;
        tail->pNext = pHead;

        delete originalHead;
    }



    void RemoveLast()
    {
        if (IsEmpty())
        {
            cout << "The List is empty." << endl;
            return;
        }

        if (pHead->pNext == pHead)
        {
            RemoveFirst();
            return;
        }

        Node* tail = pHead->pPrev;
        Node* beforeTail = tail->pPrev;
        pHead->pPrev = beforeTail;
        beforeTail->pNext = pHead;
        delete tail;
    }

    void RemoveAfter(Node* node) {
        if (node == nullptr) {
            cout << "Node removed can not be null" << endl;
            return;
        }

        if (pHead == node && pHead->pNext == node)
        { 
            cout << "Can not remove after the last node of the list." << endl;
            return;
        }

        Node* temp = node->pNext;
        if (temp == pHead) 
        {
            cout << "Can not remove after the last node of the list." << endl;
            return;
        }
        node->pNext = temp->pNext;
        temp->pNext->pPrev = node;
        delete temp;
    }

    Node* SearchNode(int value) {
        if (IsEmpty()) {
            cout << "The list is empty" << endl;
            return nullptr;
        }

        Node* current = pHead;
        do {
            if (current->info == value) {
                return current;
            }
            current = current->pNext;
        } while (current != pHead);

        cout << "Cannot find a node with " << value << endl;
        return nullptr;
    }
    void RemoveNode(Node* node) {
        if (node == nullptr) return;
        if (node == pHead) {
            if (pHead->pNext == pHead) {
                delete pHead;
                pHead = nullptr;
            }
            else {
                Node* temp = pHead;
                while (temp->pNext != pHead) {
                    temp = temp->pNext;
                }
                temp->pNext = pHead->pNext;
                delete pHead;
                pHead = temp->pNext;
            }
        }
        else {
            Node* temp = pHead;
            while (temp->pNext != node) {
                temp = temp->pNext;
            }
            temp->pNext = node->pNext;
            delete node;
        }
    }

    list<Node*> SearchNode(bool (*criteria)(Node*)) {
        list<Node*> result;
        if (IsEmpty()) {
            cout << "The list is empty" << endl;
            return result;
        }

        Node* current = pHead;
        do {
            if (criteria(current)) {
                result.push_back(current);
            }
            current = current->pNext;
        } while (current != pHead);

        return result;
    }

    static bool IsMoreThan5(Node* node) {
        return node->info > 5;
    }

    static bool IsSquared(Node* node) {
        int root = sqrt(node->info);
        return root * root == node->info;
    }

    void PrintList() {
        if (IsEmpty()) {
            cout << "The list is empty" << endl;
            return;
        }

        Node* current = pHead;
        do {
            cout << current->info << " ";
            current = current->pNext;
        } while (current != pHead);
        cout << endl;
    }

    void SelectionSort() {
        if (IsEmpty() || pHead->pNext == pHead) {
            return;
        }

        Node* sortedHead = nullptr;

        while (pHead != nullptr) {
            Node* min = pHead;
            Node* current = pHead->pNext;

            do {
                if (current->info < min->info) {
                    min = current;
                }
                current = current->pNext;
            } while (current != pHead);

            if (min->pNext == min && min->pPrev == min) {
                pHead = nullptr;
            }
            else {
                min->pPrev->pNext = min->pNext;
                min->pNext->pPrev = min->pPrev;

                if (min == pHead) {
                    pHead = min->pNext;
                }
            }

            if (sortedHead == nullptr) {
                sortedHead = min;
                min->pNext = min->pPrev = min;
            }
            else {
                Node* last = sortedHead->pPrev;
                last->pNext = min;
                min->pPrev = last;
                min->pNext = sortedHead;
                sortedHead->pPrev = min;
            }
        }

        pHead = sortedHead;
    }

    int distance(Node* p) {
        int count = 0;
        Node* current = pHead;
        while (current != p) {
            current = current->pNext;
            count++;
        }
        return count;
    }

    Node* back(int step) {
        Node* runner = pHead;
        for (int i = 0; i < step; i++) {
            runner = runner->pNext;
        }
        return runner;
    }

    Node* Partition(Node* low, Node* high) {
        Node* pivot = high;
        Node* i = low->pPrev;

        Node* j = low;
        while (j != high) {
            if (j->info < pivot->info) {
                i = (i == nullptr) ? low : i->pNext;
                swap(i, j);
            }
            j = j->pNext;
        }

        i = (i == nullptr) ? low : i->pNext;
        swap(i, high);

        return i;
    }


    void QuickSortRecursive(Node* low, Node* high) {
        if (low == nullptr || high == nullptr || low == high || low == high->pPrev) {
            return;
        }

        Node* pivot = Partition(low, high);

        if (pivot != low) {
            QuickSortRecursive(low, pivot->pPrev);
        }
        if (pivot != high) {
            QuickSortRecursive(pivot->pNext, high);
        }
    }

    void QuickSort() {
        if (pHead == nullptr) return;

        Node* tail = pHead->pPrev;
        QuickSortRecursive(pHead, tail);
    }

    void MergeLists(CircularDoublyLinkedList& list2) {
        if (IsEmpty()) {
            pHead = list2.pHead;
            return;
        }

        if (list2.IsEmpty()) {
            return;
        }

        Node* tail1 = pHead->pPrev;
        Node* tail2 = list2.pHead->pPrev;

        tail1->pNext = list2.pHead;
        list2.pHead->pPrev = tail1;

        tail2->pNext = pHead;
        pHead->pPrev = tail2;
    }

    void RemoveAllNodesByCriteria(bool (*criteria)(Node*)) {
        if (IsEmpty()) {
            cout << "The list is empty." << endl;
            return;
        }

        Node* current = pHead;
        Node* nextNode = nullptr;
        Node* prevNode = nullptr;

        do {
            nextNode = current->pNext;
            if (criteria(current)) {
                if (current == pHead) {
                    pHead = nextNode;
                    if (pHead == current) {
                        pHead = nullptr;
                    }
                }

                RemoveNode(current);

                if (IsEmpty()) {
                    break;
                }
            }
            else {
                prevNode = current;
            }

            current = nextNode;
        } while (current != pHead);

        if (!IsEmpty() && prevNode) {
            prevNode->pNext = pHead;
        }
    }


    void ReverseList() {
        if (IsEmpty() || pHead->pNext == pHead) {
            return;
        }

        Node* current = pHead;
        do {
            Node* temp = current->pNext;
            current->pNext = current->pPrev;
            current->pPrev = temp;

            current = current->pPrev;
        } while (current != pHead);

        pHead = pHead->pPrev;
    }
};

int main() 
{
    CircularDoublyLinkedList list1;
    CircularDoublyLinkedList list2;
    list1.AddFirst(10);
    list1.AddLast(20);
    list1.AddLast(30);
    list1.AddFirst(5);
    list2.AddFirst(15);
    list2.AddLast(25);
    list2.AddLast(35);

    cout << "List 1 after adding elements: ";
    list1.PrintList();
    cout << "List 2 after adding elements: ";
    list2.PrintList();

    list1.MergeLists(list2);
    cout << "List 1 after merging with List 2: ";
    list1.PrintList();

    list1.ReverseList();
    cout << "List 1 after reversing: ";
    list1.PrintList();

    list1.SelectionSort();
    cout << "List 1 after selection sort: ";
    list1.PrintList();

    list1.QuickSort();
    cout << "List 1 after quick sort: ";
    list1.PrintList();

    CircularDoublyLinkedList::Node* node = list1.SearchNode(20);
    if (node) {
        cout << "Found node with value: " << node->info << endl;
    }

    list1.RemoveNode(node);
    cout << "List 1 after removing node with value 20: ";
    list1.PrintList();

    node = list1.SearchNode(30);
    if (node) {
        list1.RemoveAfter(node);
        cout << "List 1 after removing node after 30: ";
        list1.PrintList();
    }

    list1.RemoveFirst();
    cout << "List 1 after removing the first node: ";
    list1.PrintList();

    list1.RemoveLast();
    cout << "List 1 after removing the last node: ";
    list1.PrintList();

    list1.RemoveAllNodesByCriteria(list1.IsMoreThan5);
    cout << "List 1 after removing nodes greater than 5: ";
    list1.PrintList();

    list1.RemoveAllNodesByCriteria(list1.IsSquared);
    cout << "List 1 after removing squared nodes: ";
    list1.PrintList();

    list1.RemoveFirst();

    list1.AddFirst(10);
    list1.AddLast(20);
    list1.AddLast(30);
    list1.AddFirst(5);
    cout << " Reconstruct list 1: ";
    list1.PrintList();

    node = list1.SearchNode(20);
    if (node) {
        list1.RemoveAfter(node);
        cout << "List 1 after removing node after 20: ";
        list1.PrintList();
    }

    list1.RemoveFirst();
    cout << "List 1 after removing the first node: ";
    list1.PrintList();

    list1.RemoveLast();
    cout << "List 1 after removing the last node: ";
    list1.PrintList();

    list1.RemoveAllNodesByCriteria(list1.IsSquared);
    cout << "List 1 after removing squared nodes: ";
    list1.PrintList();

    return 0;
}

File: Nhóm 11_Trần Hữu Đức_Chung Trường Vũ_23110018_23110072/File Nộp Bài/bai1a_DSA_c#/CircularLinkedListOperations.cs/CircularLinkedList.cs
using System;
using System.CodeDom;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Runtime.Remoting.Messaging;
using System.Text;
using System.Threading.Tasks;

namespace CircularLinkedListOperations.cs
{
    public class CircularLinkedList
    {
        private Node head;

        // Constructor:
        public CircularLinkedList()
        {
            head = null;
        }
        //  Initalize a list
        public void InitializeList()
        {
            Console.WriteLine("Enter the number of products: ");
            int n;
            while (!int.TryParse(Console.ReadLine(), out n) || n <= 0)
            {
                Console.WriteLine("Invalid input. Please enter a positive number: ");
            }

            for (int i = 0; i < n; i++)
            {
                Console.WriteLine($"Enter details for product {i + 1}:");

                Console.Write("Product name: ");
                string name = Console.ReadLine();

                float price;
                Console.Write("Product price: ");
                while (!float.TryParse(Console.ReadLine(), out price) || price < 0)
                {
                    Console.WriteLine("Invalid price. Please enter a non-negative number: ");
                }

                Product product = new Product(name, price);
                AddLast(CreateNode(product)); // Add product to the end of the list
            }

            Console.WriteLine("Product list has been initialized.");
        }


        // Check whether a node is empty or not
        public bool IsEmpty()
        {
            return head == null;
        }

        // Create a new node
        public Node CreateNode(Product product)
        {
            return new Node(product);
        }

        // Insert a new node at the beginning of the list
        public void AddFirst(Node newNode)
        {

            if (IsEmpty())
            {
                head = newNode;
                head.Next = head;
                head.Prev = head;
            }
            else
            {
                Node tail = head.Prev;
                newNode.Next = head;
                newNode.Prev = tail;
                head.Prev = newNode;
                tail.Next = newNode;
                head = newNode; // the difference
            }
        }

        // Insert a new node at the end of the list
        public void AddLast(Node newNode)
        {

            if (IsEmpty())
            {
                head = newNode;
                head.Next = head;
                head.Prev = head;
            }
            else
            {
                Node tail = head.Prev;
                newNode.Next = head;
                newNode.Prev = tail;
                tail.Next = newNode;
                head.Prev = newNode;
            }
        }

        // Insert a new node after a node of the list
        public void AddAfter(Node newNode, string afterProductName)
        {
            if (IsEmpty()) return;

            Node temp = head;
            do
            {
                if (temp.Data.Name == afterProductName)
                {
                    Node right = temp.Next;
                    temp.Next = newNode;
                    right.Prev = newNode;
                    newNode.Prev = temp;
                    newNode.Next = right;
                    return;
                }
                temp = temp.Next;
            } while (temp != head);

            Console.WriteLine("The node to insert after was not found");
        }

        public void RemoveFirst()
        {
            if (IsEmpty()) return;

            if (head.Next == head)
            {

                head = null;
            }
            else
            {
                Node tail = head.Prev;
                head = head.Next;
                head.Prev = tail;
                tail.Next = head;
            }
        }
        public void RemoveLast()
        {
            if (IsEmpty()) return;

            if (head.Next == head)
            {
                head = null;
            }
            else
            {
                Node tail = head.Prev;
                Node newTail = tail.Prev;
                newTail.Next = head;
                head.Prev = newTail;
            }
        }
        //Remove a node after a node of the list
        public void RemoveAfter(string afterProductName)
        {
            Node temp = head;
            do
            {
                if (temp.Data.Name == afterProductName)
                {
                    Node nodeToRemove = temp.Next;

                    if (nodeToRemove == head) return; // doen not exist node after specific node

                    temp.Next = nodeToRemove.Next;
                    nodeToRemove.Next.Prev = temp;
                    return;
                }
                temp = temp.Next;
            } while (temp != head);

            Console.WriteLine("The node to delete after was not found.");
        }
        // removenode
        public void RemoveNode(Node nodeToRemove)
        {
            if (IsEmpty() || nodeToRemove == null) return;


            if (nodeToRemove == head)
            {
                RemoveFirst();
            }

            else if (nodeToRemove == head.Prev)
            {
                RemoveLast();
            }
            else
            {
                nodeToRemove.Prev.Next = nodeToRemove.Next;
                nodeToRemove.Next.Prev = nodeToRemove.Prev;
            }
        }

        // Search a node by product name
        public Node SearchNode(string productName)
        {
            if (IsEmpty())
            {
                Console.WriteLine("the list is empty");
                return null;
            }

            Node temp = head;
            do
            {
                if (temp.Data.Name == productName)
                {
                    return temp;
                }
                temp = temp.Next;
            } while (temp != head);

            Console.WriteLine($"No product with the specified name was found.: {productName}");
            return null;
        }

        // Search nodes in the list satisfying a specific criteria. 
        public List<Node> SearchNodesByCriteria(Func<Product, bool> criteria) //Refer to Func<Product, bool> criteria from GPT.
        {
            List<Node> result = new List<Node>();
            Node temp = head;
            do
            {
                if (criteria(temp.Data))
                {
                    result.Add(temp);
                }
                temp = temp.Next;
            } while (temp != head);

            return result;
        }

        // Print the list
        public void PrintList()
        {
            if (IsEmpty())
            {
                Console.WriteLine("the list is empty");
                return;
            }

            Node temp = head;
            do
            {
                Console.WriteLine($"name: {temp.Data.Name}, price: {temp.Data.Price}");
                temp = temp.Next;
            } while (temp != head);
        }
        // sort the list------------------------------------------------------------------------------------------------------------
        // declare
        public int CompareByPrice(Product p1, Product p2)
        {
            return p2.Price.CompareTo(p1.Price);
        }
        public int CompareByName(Product p1, Product p2)
        {
                return string.Compare(p1.Name, p2.Name);
        }
        // selection sort
        public void SelectionSort()
        {
            if (head == null) return;

            CircularLinkedList sortedList = new CircularLinkedList();

            while (head != null)
            {
                Node min = head;
                Node current = head.Next;

                // find the min node
                while (current != head)
                {
                    if (compareMethod(min.Data, current.Data) > 0)
                    {
                        min = current;
                    }
                    current = current.Next;
                }
                // throw this node to other list
                RemoveNode(min);
                sortedList.AddFirst(min);
            }
            // update
            this.head = sortedList.head;
        }
        // quicksort
        private int distance (Node p)
        {
            int count = 0;
            for (Node i = head; i != p; i = i.Next) count++;
            return count;
        }

        private Node back ( int step)
        {
            Node runner = head;
            for (int i = 1; i <=step; i++) runner = runner.Next;
            return runner;
        }
        private  Node Partition(Node low, Node high)
        {
            Node pivot = high;
            Node i = low;
            while (i != pivot)
            {
                if  (i.Data.Price < pivot.Data.Price)
                {
                    Node j = i.Next;
                    RemoveNode(i);
                    AddAfter(i, pivot.Data.Name);
                    i = j;
                }
                else i = i.Next;
            }

            return pivot;
        }

        public void QuickSortRecursive(Node low, Node high)
        {

            if (low != high) 
            {
                if (low.Next == high && low.Data.Price > high.Data.Price) return; // loại trường hợp kề đúng vị trí 4 3
                int dislow = distance(low);
                int dishigh = distance(high);
                Node pivot = Partition(low, high);
                low = back(dislow);
                high = back(dishigh);
                if (pivot != high)
                {
                    QuickSortRecursive(low, pivot);
                    QuickSortRecursive(pivot.Next, high);
                }
                else
                {
                    pivot = pivot.Prev;
                    QuickSortRecursive(low, pivot);
                    QuickSortRecursive(pivot.Next, high);
                }
            }
        }
        public  void QuickSort()
        {

            if (head == null) return;

            Node tail = head.Prev;
            QuickSortRecursive(head, tail);
        }

        // 3 operations -------------------------------------------------------------------------------------------------------------
        // Merge 2 lists
        public void MergeSortedLists(CircularLinkedList otherList)
        {
            if (otherList.IsEmpty()) return; // other list is empty

            if (this.IsEmpty())// this list is empty
            {
                this.head = otherList.head;
                return;
            }
            // 2 list is not empty
            Node tail = this.head.Prev;
            Node othertail = otherList.head.Prev;

            tail.Next = otherList.head;
            otherList.head.Prev = tail;
            othertail.Next = head;
            head.Prev = othertail;
        }
        // remove all nodes satisfying a specific criteria
        public void RemoveNodesByCondition(Func<Product, bool> condition)
        {
            if (IsEmpty()) return;
            CircularLinkedList otherList = new CircularLinkedList();
            Node current = head;

            do
            {
                if (condition(current.Data))
                {
                    
                    current = current.Next;
                }
                else
                {
                    
                    Node tmp = current.Next;
                    RemoveNode(current);
                    otherList.AddFirst(current);
                    current = tmp;
                }
            } while (!IsEmpty() && current != head);
            head = otherList.head;
        }



        // Reverse the list
        public void ReverseList()
        {
            CircularLinkedList otherList = new CircularLinkedList();
            if (IsEmpty()) return;
            Node runner = head;
            do 
            {
                Node tmp = runner.Next;
                RemoveNode(runner);
                otherList.AddFirst(runner);
                runner = tmp;
            }
            while (!IsEmpty());
            head = otherList.head;
        }


    }
}